Mathlib/Tactic/PushNeg.lean

```diff
@@ -35,6 +36,12 @@ theorem not_lt_eq (a b : β) : (¬ (a < b)) = (b ≤ a) := propext not_lt
 theorem not_ge_eq (a b : β) : (¬ (a ≥ b)) = (a < b) := propext not_le
 theorem not_gt_eq (a b : β) : (¬ (a > b)) = (a ≤ b) := propext not_lt
 
+variable {γ : Type v}
+theorem not_nonempty_eq (s : Set γ) : (¬ s.Nonempty) = (s = ∅) := by
```

Comment:
Why can't we use the existing `not_nonempty_iff_eq_empty` lemma here?

Reply:
Because of the import order: `not_nonempty_iff_eq_empty` is in a file which imports `PushNeg`. Instead of trying to shuffle everything around, I moved as little as possible to the file `Init.Set` (only the definition of `Set.Nonempty`) and went for a direct proof here.

Reply:
Should we just move the existig lemma then, since it has basically the same statement?

Reply:
The existing lemma has a natural one-liner proof, while the one here is unnatural because the theory hasn't been built yet. I'm more thinking of the lemma `not_nonempty_eq` as an artifact which is necessary for `push_neg` but doesn't really take part in theory building, and keep things as they are currently.

Reply:
Can we instead add some kind of `push_neg` extension so that this can go in an appropriate file? I feel like we'll want that anyway to add the analogous behavior for `not (Finset.Nonempty s)`, but likely won't have the luxury of being able to import it at all...

Reply:
`Not (Nontrivial X)` and `Not (Nonempty X)` seem like other good candidates, which may or may not be easy to add into this file without an extension mechanism.

Reply:
An extension mechanism would definitely be even better, but this is out of my pay grade: the goal of the PR is just for this single case that I've encountered several times before, while the other ones have never bitten me.

Reply:
I think it's nice when meta code states its own theorems, since that protects itself from future changes to mathlib (this is a place where we otherwise rely entirely on tests -- the meta code doesn't have any compile-time typechecking for referenced theorems).

Regarding extension mechanisms, I don't think it's worth it if we're only going to add a handful of things to `push_neg`. The cost of maintaining an extension mechanism is significantly higher than adding extra cases in the code, even if it involves duplicating a bit of code (and this is only three lines, so I think we're doing alright).

@sgouezel Could you remove the `variable {γ : Type v}`? We have `autoImplicits` on in this file so you don't need it.